oc: Use default resources where it makes sense

kargakis opened the pull request:
Commands I have defaulted using the builder/RESTMapper in this PR:
* oc deploy 
(Was using both syntaxes before, now achieves it via the resource builder.)

* oadm build-chain 
(Was using both syntaxes before, now achieves it via the RESTMapper.)

* oc start-build 
(Was using only NAME syntax, now uses bc/NAME too.)

* oadm secrets add 
(Was using only TYPE/NAME syntax, now uses NAME too.)

* oc cancel-build 
(Was using only NAME syntax, now uses build/NAME too.)

@smarterclayton @fabianofranz 

Fixes https://github.com/openshift/origin/issues/4781
Fixes https://github.com/openshift/origin/issues/1995

deads2k commented on pkg/cmd/cli/cmd/deploy.go:
```diff
@@ -174,7 +170,7 @@ func (o *DeployOptions) Validate(args []string) error {
 	r := o.builder.
 		NamespaceParam(o.namespace).
-		ResourceTypeOrNameArgs(false, o.deploymentConfigName).
+		ResourceNames("deploymentConfigs", o.deploymentConfigName).
```
lowercase?

deads2k commented:
This command is requiring resource/name format: https://github.com/openshift/origin/blob/master/pkg/cmd/cli/secrets/add_secret_to_obj.go

liggitt commented:
> This command is requiring resource/name format: https://github.com/openshift/origin/blob/master/pkg/cmd/cli/secrets/add_secret_to_obj.go

For two args, no less

kargakis commented:
[test]

kargakis commented:
@deads2k jenkins is green, merge at your will

liggitt commented:
Are there tests to make sure prior invocations still work correctly?

kargakis commented:
> Are there tests to make sure prior invocations still work correctly?

`test-cmd.sh` should catch such errors.

liggitt commented:
I agree it should, just not sure it does :)
Can you summarize what invocations were allowed prior, and what are allowed now?
For example, abbreviations like `oc secrets add sa/sa-name ...` were allowed, but I don't think test-cmd.sh exercises that

kargakis commented:
> Can you summarize what invocations were allowed prior (with/without types, abbreviations and full resource names), and what are allowed now?

I've updated the first post with more details about and I've also updated test-cmd to test all cases.

smarterclayton commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -102,10 +102,31 @@ func RunStartBuild(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args
-	if len(name) == 0 {
-		name = args[0]
+	if len(name) == 0 && len(args[0]) > 0 {
+		parts := strings.Split(args[0], "/")
```
This seems like something that should be handled by resource builder?

kargakis replied:
I am not using the resource builder here because it would try to get the buildConfig (which we are not interested in) in order to start the build. Instead, I think using the RESTMapper to resolve the resource type is enough.

smarterclayton replied:
It makes me believe that what you are representing here is a common pattern
that should be abstracted either into resource.Builder (but without the
fetch) or as supporting utility code in the cmd packages.

On Oct 11, 2015, at 3:38 PM, Michail Kargakis <notifications@github.com>
wrote:

In pkg/cmd/cli/cmd/startbuild.go
<https://github.com/openshift/origin/pull/4947#discussion_r41714535>:

>  	name := buildName
>  	isBuild := true
> -	if len(name) == 0 {
> -		name = args[0]
> +	if len(name) == 0 && len(args[0]) > 0 {
> +		parts := strings.Split(args[0], "/")

I am not using the resource builder here because it would try to get the
buildConfig (which we are not interested in) in order to start the build.
Instead, I think using the RESTMapper to resolve the resource type is
enough.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4947/files#r41714535>.

kargakis replied:
Added an utility under origin/pkg/cmd/util to handle this.

smarterclayton commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -94,12 +93,29 @@ func (o *BuildChainOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, a
-	o.name, o.tag, err = buildChainInput(args[0])
-	if err != nil {
-		return cmdutil.UsageError(cmd, err.Error())
+	parts := strings.Split(args[0], "/")
```
This seems like it changes the behavior of the lookup "namespace/name:tag" for ists.  So this breaks existing usage?

smarterclayton replied:
Ah, build chain uses resource syntax.

Still, the duplication of this code seems high.  Why is split handled outside of resource.Builder?

kargakis replied:
The resource builder will try to construct the istag, but if that doesn't exist *currently*, we will get an error. Instead, by using the RESTMapper to just resolve the syntax, we let the graph code to create graphs for tags that are outputs for builds (istags that aren't created yet).

deads2k commented on pkg/cmd/util/cmd.go:
```diff
@@ -31,3 +33,32 @@ func GetDisplayFilename(filename string) string {
+	parts := strings.Split(arg, "/")
+	switch len(parts) {
+	case 1:
+		name = parts[0]
```
Shouldn't the default kind be applied for cases where you only have a name?

kargakis replied:
What you are asking about is `ResourceNames` in the resource builder. This is different.

deads2k commented on pkg/cmd/util/cmd.go:
```diff
@@ -31,3 +33,32 @@ func GetDisplayFilename(filename string) string {
+// ResolveResourceSyntax will return the kind (if found), name and a possible error
+// from the provided arg. If a default kind has been specified, then arg
+// should respect it.
+func ResolveResourceSyntax(mapper meta.RESTMapper, defaultKind, arg string) (string, string, error) {
```
I want this kind of resolution upstream too.  This is would allow resolution of types for logs without retrieving resources I don't care about.

deads2k commented on pkg/cmd/util/cmd.go:
```diff
@@ -31,3 +33,32 @@ func GetDisplayFilename(filename string) string {
+		if err != nil {
+			return "", "", err
+		}
+		if len(defaultKind) > 0 && kind != defaultKind {
```
This looks like weird behavior for a field called `defaultKind`.  I'd rather have a `defaultKind` than a `requiredKind`.

kargakis replied:
I've renamed to `requiredKind` because that's what its' supposed to be. This function is supposed to be called instead of the resource builder when we don't want to fetch the object but it shouldn't default to a resource, and allow others as well (how `ResourceNames` works), it should enforce the resource if it's specified (if not, then we just preserve `oc cmd NAME` syntax).

deads2k commented on pkg/cmd/util/cmd.go:
```diff
@@ -31,3 +33,32 @@ func GetDisplayFilename(filename string) string {
+// ResolveResourceSyntax will return the kind (if found), name and a possible error
+// from the provided arg. If a required kind has been specified, then arg
+// should respect it.
+func ResolveResourceSyntax(mapper meta.RESTMapper, requiredKind, arg string) (string, string, error) {
```
I think this would be more obvious and more easily re-used if factored like

```go
// ResolveResource returns the resource, name, and kind of the resourceString
func ResolveResource(resourceString, defaultResource string, mapper meta.RESTMapper) (string, string, string){
}
...
In other code
if _, _, kind := ResolveResource(arg[0], "buildconfigs", mapper); kind != expectedKind{
    // fail
}
```

This factoring would allow the function to parse arguments for functions that require it, determine kinds using the mapper without branches, and make expectations obvious in the code that cares.  It's a little weird to have a `Resolve` function that forces kind matching.

kargakis replied:
ok, updated

kargakis commented:
re[test]

kargakis commented:
go stdlib race

rere[test]

kargakis commented:
Tagging more eyes:

@ironcladlou about deploy
@mfojtik about {start,cancel}-build

deads2k commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -94,12 +94,20 @@ func (o *BuildChainOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, a
-	o.name, o.tag, err = buildChainInput(args[0])
+	kind, defaultKind := "", "ImageStreamTag"
+	mapper, _ := f.Object()
+	kind, o.name, err = osutil.ResolveResource(defaultKind, args[0], mapper)
```
I think what you really want is a way to get a canonical resource representation.  Something where you get "imagestreamtag" back for "istag", "istags", "imagestreamtags" in any case.  If multiple resources deal in the same kind, this breaks because "notAnIsTag/foo" would match.

I asked @smarterclayton if he knew whether it existed, he said he'd look for it/make it.

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -102,10 +103,27 @@ func RunStartBuild(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args
+	var (
+		name        = buildName
+		kind        = ""
+		defaultKind = "BuildConfig"
```
should be using resources:
```
defaultResource = "buildconfigs"
```

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -102,10 +103,27 @@ func RunStartBuild(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args
+
+	if len(name) == 0 && len(args[0]) > 0 {
+		mapper, _ := f.Object()
+		kind, name, err = osutil.ResolveResource(defaultKind, args[0], mapper)
```
resource, name, err = ...

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -102,10 +103,27 @@ func RunStartBuild(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args
+		if err != nil {
+			return err
+		}
+		if kind != defaultKind {
```
this check is hard to follow, I'd rather see a switch case below like 

switch resource {
case "buildconfigs":
  ... handle build configs ...
default:
  return fmt.Errorf("invalid resource provided: %s", resource)
}

kargakis replied:
We only accept buildConfigs as arguments for start-build, (builds are passed through --from-build) so this switch is:
```go
switch resource {
case "buildconfigs":
default:
  return fmt.Errorf("invalid resource provided: %s", resource)
}
```
Now how is this better than
```go
if resource != "buildconfigs" {
	return fmt.Errorf("invalid resource provided: %s", resource)
}
```

liggitt replied:
`start-build build/a` should be allowed if we allow starting builds from builds

kargakis replied:
> or if we don't want to do that, we should handle the "builds" case specifically by telling them to use --from-build=..., rather than "invalid resource provided"

That's a good suggestion

liggitt commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -94,12 +94,20 @@ func (o *BuildChainOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, a
 
-	// Parse user input
-	o.name, o.tag, err = buildChainInput(args[0])
+	kind, defaultKind := "", "ImageStreamTag"
```
resource, defaultResource := "", "imagestreamtags"

liggitt commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -94,12 +94,20 @@ func (o *BuildChainOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, a
-		return cmdutil.UsageError(cmd, err.Error())
+		return err
+	}
+	if kind != defaultKind {
```
change to switch to make it obvious what resources we support, and make it straightforward to add support for new resources
```
switch resource {
case "imagestreamtags":
  ... handle
default:
  ... error
}
```

liggitt commented on pkg/cmd/util/cmd.go:
```diff
@@ -31,3 +33,29 @@ func GetDisplayFilename(filename string) string {
+
+// ResolveResource returns the resource name and kind of the resourceString.
+// If the resource string has no specified kind, defaultKind will be returned.
+func ResolveResource(defaultKind, resourceString string, mapper meta.RESTMapper) (string, string, error) {
```
defaultResource

liggitt commented on pkg/cmd/util/cmd.go:
```diff
@@ -31,3 +33,29 @@ func GetDisplayFilename(filename string) string {
 	return filename
 }
+
+// ResolveResource returns the resource name and kind of the resourceString.
```
return resource and name

liggitt commented on pkg/cmd/util/cmd.go:
```diff
@@ -31,3 +33,29 @@ func GetDisplayFilename(filename string) string {
+	case 1:
+		name = parts[0]
+	case 2:
+		_, kind, err := mapper.VersionAndKindForResource(parts[0])
```
I think this validates the resource is known, and handles expanding short forms to long forms, which is good

liggitt commented on pkg/cmd/util/cmd.go:
```diff
@@ -31,3 +33,29 @@ func GetDisplayFilename(filename string) string {
+			return "", "", err
+		}
+		name = parts[1]
+		return kind, name, nil
```
once we know the resource is good, and have the canonical Kind, I *think* we want to convert it back to resource form to return it using `KindToResource`

kargakis replied:
Do we handle subresources in the cli anywhere else? The resource builder doesn't for sure...

liggitt replied:
right, just trying not to set up things that will break in the future

kargakis commented:
@liggitt @deads2k comments addressed

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -102,10 +103,30 @@ func RunStartBuild(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args
+	var (
+		name     = buildName
+		resource = ""
+		isBuild  = true
```
having both `isBuild` and `resource` is confusing... seems like this should start out as
```
name = buildName
resource = "builds"
```

and the code below that switches on isBuild should switch on resource

kargakis replied:
ok

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -102,10 +103,30 @@ func RunStartBuild(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args
+			return err
+		}
+		switch resource {
+		case "buildconfigs", "buildconfig":
```
if ResolveResource normalizes, which it should, you only have to handle the plural case

kargakis replied:
ok

liggitt commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -94,12 +94,24 @@ func (o *BuildChainOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, a
+	}
+
+	switch resource {
+	case "imagestreamtags", "imagestreamtag":
```
should only need to handle the plural case

kargakis replied:
ok

liggitt commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -94,12 +94,24 @@ func (o *BuildChainOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, a
 	}
-	glog.V(4).Infof("Using '%s:%s' as the image stream tag to look dependencies for", o.name, o.tag)
+
+	if !strings.Contains(o.name, ":") {
```
Move this into the `case "imagestreamtags":` above, since this only makes sense for imagestreamtag resources. If we add support for another kind in the future, we wouldn't want to automatically pick up this behavior for it

kargakis replied:
ok

liggitt commented on pkg/cmd/experimental/buildchain/buildchain.go:
```diff
@@ -94,12 +94,24 @@ func (o *BuildChainOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, a
+
+	if !strings.Contains(o.name, ":") {
+		// Default to latest if the user haven't specified
+		o.name = o.name + ":" + imageapi.DefaultImageTag
```
this feels like it should be a peer method to SplitImageStreamTag/JoinImageStreamTag, like NormalizeImageStreamTag

kargakis replied:
added NormalizeImageStreamTag

liggitt commented on pkg/cmd/util/cmd.go:
```diff
@@ -31,3 +33,30 @@ func GetDisplayFilename(filename string) string {
 	return filename
 }
+
+// ResolveResource returns the resource name and resource of the resourceString.
```
godoc is backwards of what is actually returned

kargakis replied:
fixed

liggitt commented on test/cmd/admin.sh:
```diff
@@ -160,6 +160,9 @@ echo "registry: ok"
 oc process -f examples/sample-app/application-template-stibuild.json -l build=sti | oc create -f -
+# Test both the type/name resource syntax and the fact that istag/origin-ruby-sample:latest is still
+# not created but due to a buildConfig pointing to it, we get back its graph of deps.
+[ "$(oadm build-chain istag/origin-ruby-sample | grep 'imagestreamtag/origin-ruby-sample:latest')" ]
```
need a test to make sure `--from-build` doesn't break

kargakis replied:
Added. I was surprised we didn't have one already...

liggitt commented on pkg/cmd/util/cmd.go:
```diff
@@ -31,3 +33,30 @@ func GetDisplayFilename(filename string) string {
+	case 1:
+		name = parts[0]
+	case 2:
+		_, resource, err := mapper.VersionAndKindForResource(parts[0])
```
call this `kind` so we don't get confused

kargakis replied:
ok

On Fri, Oct 16, 2015 at 4:35 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/cmd/util/cmd.go
> <https://github.com/openshift/origin/pull/4947#discussion_r42249118>:
>
> > @@ -31,3 +33,30 @@ func GetDisplayFilename(filename string) string {
> >
> >  	return filename
> >  }
> > +
> > +// ResolveResource returns the resource type and name of the resourceString.
> > +// If the resource string has no specified type, defaultResource will be returned.
> > +func ResolveResource(defaultResource, resourceString string, mapper meta.RESTMapper) (string, string, error) {
> > +	if mapper == nil {
> > +		return "", "", errors.New("mapper cannot be nil")
> > +	}
> > +
> > +	var name string
> > +	parts := strings.Split(resourceString, "/")
> > +	switch len(parts) {
> > +	case 1:
> > +		name = parts[0]
> > +	case 2:
> > +		_, resource, err := mapper.VersionAndKindForResource(parts[0])
>
> call this kind so we don't get confused
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4947/files#r42249118>.
>

kargakis commented:
Is there any other comment here? @smarterclayton @liggitt @deads2k 

@mfojtik @bparees

mfojtik commented:
LGTM (start-build / cancel-build)

liggitt commented on pkg/cmd/cli/cmd/deploy.go:
```diff
@@ -138,11 +138,7 @@ func (o *DeployOptions) Complete(f *clientcmd.Factory, args []string, out io.Wri
 	o.out = out
 
 	if len(args) > 0 {
```
are we discarding args if they give us more than 1? should we error on that?

kargakis replied:
We already do; see https://github.com/openshift/origin/blob/c4c01507f6d1222e3284835081295b6ff030ebf5/pkg/cmd/cli/cmd/deploy.go#L121-L123

liggitt replied:
oh, ha

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -102,23 +103,36 @@ func RunStartBuild(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args
+			return err
+		}
+		switch resource {
+		case "buildconfigs":
```
add a `// no special handling required` comment to make it clear this isn't intended to be a fall-through case

kargakis replied:
ok

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -135,11 +149,12 @@ func RunStartBuild(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args
+	case "buildconfigs":
 		if newBuild, err = client.BuildConfigs(namespace).Instantiate(request); err != nil {
 			return err
 		}
```
add a `default` case that errors

kargakis replied:
We already have that covered in L126, but uf you really want to check it here too, I can add it.

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -210,7 +225,7 @@ func RunListBuildWebHooks(f *clientcmd.Factory, out, errOut io.Writer, name stri
 	}
 
-	if isBuild {
+	if resource == "builds" {
```
switch on resource and handle buildconfigs, builds, and default error. In general, I want to make sure we don't make assumptions about resources for when we want to accept other resources

kargakis replied:
ok

liggitt commented on pkg/cmd/util/cmd.go:
```diff
@@ -31,3 +33,30 @@ func GetDisplayFilename(filename string) string {
+
+// ResolveResource returns the resource type and name of the resourceString.
+// If the resource string has no specified type, defaultResource will be returned.
+func ResolveResource(defaultResource, resourceString string, mapper meta.RESTMapper) (string, string, error) {
```
need lots of tests around this. Off the top of my head:
```
// invalid cases
"", "a/b/c"   -> "", "", error about unknown format
"", "foo/bar" -> "", "", error about unknown resource foo

// empty resource string cases
"",    ""    -> "",    "",    nil
"",    "bar" -> "",    "bar", nil
"foo", "bar" -> "foo", "bar", nil

// shortname, case-insensitive (preserving name case), singular/plural K8s resources from resource string
"foo","rc/bar"                     -> "replicationcontrollers", "bar", nil
"foo","replicationcontroller/bar"  -> "replicationcontrollers", "bar", nil
"foo","replicationcontrollers/bar" -> "replicationcontrollers", "bar", nil
"foo","ReplicationControllers/bar" -> "replicationcontrollers", "bar", nil
"foo","ReplicationControllers/Bar" -> "replicationcontrollers", "Bar", nil

// shortname, case-insensitive (preserving name case), singular/plural origin resources from resource string
"foo","is/bar"           -> "imagestreams", "bar", nil
"foo","imagestream/bar"  -> "imagestreams", "bar", nil
"foo","imagestream/bar"  -> "imagestreams", "bar", nil
"foo","imagestreams/bar" -> "imagestreams", "bar", nil
"foo","ImageStreams/bar" -> "imagestreams", "bar", nil
"foo","ImageStreams/Bar" -> "imagestreams", "Bar", nil
```

kargakis replied:
done

liggitt commented on pkg/image/api/helper.go:
```diff
@@ -198,6 +198,16 @@ func JoinImageStreamTag(name, tag string) string {
+// NormalizeImageStreamTag normalizes an image stream tag by defaulting to 'latest'
+// if no tag has been specified.
+func NormalizeImageStreamTag(name string) string {
+	if !strings.Contains(name, ":") {
```
reuse the existing logic:
```
name, tag, _ := SplitImageStreamTag(name)
return JoinImageStreamTag(name, tag)
```

liggitt commented on pkg/cmd/util/cmd_test.go:
```diff
@@ -0,0 +1,182 @@
+			t.Errorf("%s: expected error but got none", test.name)
+			continue
+		}
+		if !strings.EqualFold(gotResource, test.expectedResource) {
```
the expectedResource should exactly equal gotResource, so `==`, not `EqualFold`

liggitt commented on pkg/cmd/util/cmd_test.go:
```diff
@@ -0,0 +1,182 @@
+			t.Errorf("%s: expected resource type %s, got %s", test.name, test.expectedResource, gotResource)
+			continue
+		}
+		if !strings.EqualFold(gotResource, test.expectedResource) {
```
copy/paste error, compare `gotName` and `expectedName`

liggitt replied:
also, `==`, not `EqualFold`

kargakis replied:
> copy/paste error

Happens even in the best of families. Updated

kargakis commented:
If there aren't any other comments I would like this is because the resource utility will be really helpful and I won't need to use the resource builder (and thus fetch the object) [here](https://github.com/kargakis/origin/blob/82e9973aacf01426c5c23c23dd046a2f9c64ca64/pkg/cmd/cli/cmd/logs.go#L75-L86).

liggitt commented:
You can use the resource builder without actually fetching objects. Look at what the delete command does. I haven't had a chance to look at this since the last updates were made.

kargakis commented:
[test]

smarterclayton commented:
[test]

smarterclayton commented:
Test flake in https://github.com/openshift/origin/issues/5410

On Sat, Oct 24, 2015 at 6:00 PM, OpenShift Bot <notifications@github.com>
wrote:

> Origin Action Required: Pull request cannot be automatically merged,
> please rebase your branch from latest HEAD and push again
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4947#issuecomment-150854155>.
>

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -130,23 +131,37 @@ func RunStartBuild(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra
+
+	if len(name) == 0 && len(args[0]) > 0 {
+		mapper, _ := f.Object()
+		resource, name, err = osutil.ResolveResource("buildconfigs", args[0], mapper)
```
don't you need to make sure you got back a non-empty name?

kargakis replied:
ok

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -130,23 +131,37 @@ func RunStartBuild(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra
+		resource = "builds"
+	)
+
+	if len(name) == 0 && len(args[0]) > 0 {
```
the logic above that guarantees we have a single arg in this case is buried enough that I'd like an explicit `len(args) > 0` check here

kargakis replied:
ok

kargakis replied:
hm, I shouldn't remove this
```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/openshift/origin/pkg/cmd/cli/cmd.RunStartBuild(0xc20810dae0, 0x7fccbea39250, 0xc20802e000, 0x7fccbea390f0, 0xc20802e008, 0xc2082503c0, 0xc208278840, 0x0, 0x1, 0x0, ...)
	/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/pkg/cmd/cli/cmd/startbuild.go:144 +0x1817
github.com/openshift/origin/pkg/cmd/cli/cmd.func·023(0xc2082503c0, 0xc208278840, 0x0, 0x1)
	/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/pkg/cmd/cli/cmd/startbuild.go:85 +0x97
github.com/spf13/cobra.(*Command).execute(0xc2082503c0, 0xc2082787b0, 0x1, 0x1, 0x0, 0x0)
	/go/src/github.com/openshift/origin/Godeps/_workspace/src/github.com/spf13/cobra/command.go:547 +0x82f
github.com/spf13/cobra.(*Command).Execute(0xc2080a34a0, 0x0, 0x0)
	/go/src/github.com/openshift/origin/Godeps/_workspace/src/github.com/spf13/cobra/command.go:630 +0x414
main.main()
	/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/cmd/oc/oc.go:22 +0x175
```

liggitt replied:
I wasn't asking to remove that, was actually wanting this:
```
if len(name) == 0 len(args) > 0 && len(args[0]) > 0 {
  ...
}
if len(name) == 0 {
  // error
}
```

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -167,7 +182,7 @@ func RunStartBuild(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra
 	var newBuild *buildapi.Build
 	switch {
-	case !isBuild && (len(fromFile) > 0 || len(fromDir) > 0 || len(fromRepo) > 0):
+	case resource != "builds" && (len(fromFile) > 0 || len(fromDir) > 0 || len(fromRepo) > 0):
```
@smarterclayton do `fromFile`, etc, get ignored when `resource == "builds"`?

smarterclayton replied:
Should error out, will double check

On Oct 27, 2015, at 8:16 AM, Jordan Liggitt <notifications@github.com>
wrote:

In pkg/cmd/cli/cmd/startbuild.go
<https://github.com/openshift/origin/pull/4947#discussion_r43113528>:

> @@ -167,7 +182,7 @@ func RunStartBuild(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra
>
>  	var newBuild *buildapi.Build
>  	switch {
> -	case !isBuild && (len(fromFile) > 0 || len(fromDir) > 0 || len(fromRepo) > 0):
> +	case resource != "builds" && (len(fromFile) > 0 || len(fromDir) > 0 || len(fromRepo) > 0):

@smarterclayton <https://github.com/smarterclayton> do fromFile, etc, get
ignored when resource == "builds"?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4947/files#r43113528>.

liggitt commented:
one question on fromRepo, fromFile, etc interaction from @smarterclayton, then lgtm

smarterclayton commented:
`if len(args) > 0 || len(buildName) > 0 || len(fromFile) > 0 ||
len(fromDir) > 0 || len(fromRepo) > 0 {`

we're safe.

On Tue, Oct 27, 2015 at 4:32 PM, Jordan Liggitt <notifications@github.com>
wrote:

> one question on fromRepo, fromFile, etc interaction from @smarterclayton
> <https://github.com/smarterclayton>, then lgtm
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4947#issuecomment-151635360>.
>

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -129,23 +130,40 @@ func RunStartBuild(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra
+		if err != nil {
+			return err
+		}
+		if len(name) == 0 {
```
move this check outside the if block to catch cases where len(name)==0, but len(args[0]) == 0, for example

kargakis replied:
ok

liggitt commented on pkg/cmd/cli/cmd/startbuild.go:
```diff
@@ -129,23 +130,40 @@ func RunStartBuild(f *clientcmd.Factory, in io.Reader, out io.Writer, cmd *cobra
+			return err
+		}
+		if len(name) == 0 {
+			return fmt.Errorf("build config name shouldn't be empty")
```
you don't know they specified a build config... should be "resource name is required" or something

kargakis replied:
ok

liggitt commented:
LGTM, ready to merge?

kargakis commented:
Lets do it

liggitt commented:
[merge]
